nexus/src/app/sagas/instance_create.rs

```diff
@@ -299,11 +306,245 @@ impl NexusSaga for SagaInstanceCreate {
+                ActionError::action_failed(Error::internal_error(&format!(
+                    "failed to convert mac address: {e}"
+                )))
+            })?;
+
+    let vni: u32 = network_interface.vni.into();
+
+    debug!(log, "fetching external ip addresses");
+
+    let external_ips = &datastore
+        .instance_lookup_external_ips(&opctx, instance_id)
+        .await
+        .map_err(ActionError::action_failed)?;
+
+    // TODO: Correctness
+    // Do we provide a mechanism for binding specific external ip
+    // addresses to specific network interfaces? Currently we're mapping
+    // all external ip addresses to the first interface we find.
+
+    let mut new_external_ips: Vec<ExternalIp> = vec![];
+
+    for entry in external_ips {
+        debug!(log, "checking for existing nat mapping for {entry:#?}");
+
+        let existing_nat = match entry.ip {
+            ipnetwork::IpNetwork::V4(network) => {
+                dpd_client.nat_get_ipv4(&network.ip(), *entry.first_port).await
+            }
+            ipnetwork::IpNetwork::V6(network) => {
+                dpd_client.nat_get_ipv6(&network.ip(), *entry.first_port).await
+            }
+        };
+
+        match existing_nat {
+            Ok(_) => {
+                // nat entry already exists, do nothing
+            }
+            Err(_) => {
+                // TODO: correctness
+                // This needs a bit more logic since "not found" is an error,
+                // but in this case it is actually ok. We only want to bail
+                // out if a mapping is found or a more serious error occurs.
+                // Since any communication errors here will occur again in
+                // the next call to dpd, we'll do nothing for now.
+                new_external_ips.push(entry.clone());
+            }
+        }
+    }
+
+    for entry in new_external_ips {
+        debug!(log, "creating nat entry for: {entry:#?}");
+
+        let nat_target = dpd_client::types::NatTarget {
+            inner_mac: dpd_client::types::MacAddr {
+                a: mac_address.into_array().to_vec(),
+            },
+            internal_ip: *sled_ip_address.ip(),
+            vni: vni.into(),
+        };
+
+        match entry.ip {
+            ipnetwork::IpNetwork::V4(network) => {
+                dpd_client
+                    .nat_set_ipv4(
+                        &network.ip(),
+                        *entry.first_port,
+                        *entry.last_port,
+                        &nat_target,
+                    )
+                    .await
+            }
+            ipnetwork::IpNetwork::V6(network) => {
+                dpd_client
+                    .nat_set_ipv6(
+                        &network.ip(),
+                        *entry.first_port,
+                        *entry.last_port,
+                        &nat_target,
+                    )
+                    .await
+            }
+        }
+        .map_err(|e| {
```

Comment:
So I think the fact that each request can fail means these need to be separate saga actions. Look at the NIC and disk-creation for examples. We basically create N nodes, for N NICs. Each of those creates (or destroys) exactly one item. Otherwise, the action is not really idempotent.

Reply:
Thinking about this more, I'm not _certain_ this is the case. If any of the requests fail, the saga is supposed to unwind, right? Even if the node is re-attempted, any requests that were successful would not be re-sent because of the logic above this that checks for existing nat entries, so unless I'm missing something, I think this is actually idempotent.

Reply:
Persisting our chat discussion here:

I had a misunderstanding of Steno's behavior. Since we don't run `undo` actions for failed nodes, the logic in `sic_remove_network_config()` would never be called in the event of a partial failure, which is why the overall behavior would not be idempotent.